### DISCUSSION TITLE
more rotation examples / advices for text rotation and positioning

### DISCUSSION THREAD

**Discussion by abubelinha** [py-pdf/fpdf2#967] [Q&A] [answered]
Hi. I am trying to migrate to python an old php script which used fpdf to produce a  multipage pdf,.
Each page should have lateral rotated texts containing a descriptive title for each page.

As I found a similar closed issue related to text roataion, I shared an example code there: 
https://github.com/py-pdf/fpdf2/issues/190#issuecomment-1767447064

But I am not being able to achieve what I want.
Could please somebody take a look there and tell me what I am doing wrong?

Thanks a lot in advance
@abubelinha

**Comment by gmischler** [comment, marked as answer]
Welcome to fpdf2, @abubelinha!

allow me to invite you to a little thought experiment:

If you remove the rotation context from your example, where would you expect the cell text to appear on the page?
And now that you have your text somewhere, imagine it to slowly rotate around a pivot point at the very bottom of the page.
Which way will it move, and where will it end up?


Before thinking about it in this systematic way, you seem to have expected the text to be magically teleported to the pivot point.
That is evidently not how it works.

`.cell()` always places the text below and to the right of the current page position. The docstring for `rotation()` explains that this current position will not be affected by its use. Since you didn't change it either, the text wants to appear starting at the default of x=10/y=10, at the very top of the page. If you then rotate it around a pivot point 27 cm away, it will quickly go off page. What you really want here is to fist set the current position to the same location as the pivot with `.set_xy()` and use `.set_auto_page_break(False)` (because that places the text outside the page margins). Then you can continue as you did and get the desired result.

Alternatively, for placing elements in arbitrary positions on the page, you might consider using [`.FlexTemplate()`](https://py-pdf.github.io/fpdf2/Templates.html#using-flextemplate), which takes some minimal preparation, but may make the placement a bit more intuitive and, well, flexible...

**Comment by abubelinha** [comment]
Thanks a lot @gmischler 
I was already [answering myself](https://github.com/py-pdf/fpdf2/issues/190#issuecomment-1768963089) after I realized how `.cell()` works.
I solved it using `.text(x,y)` instead, but that way I couldn't set a border for the text box.

With your great suggestions I could now reuse `.cell()` again and set the border (more or less).

As I asked for more examples, here I post both methods for future reference (suggestions welcome to improve them):

```Python
from fpdf import FPDF
pdf = FPDF(orientation="P", unit="mm", format="A4")
pdf.add_page()
pdf.set_font("times")

a,x,y=90,10,295 # left margin banner using .text():
with pdf.rotation(angle=a, x=x, y=y):
banner = pdf.text(x=x, y=y,
	text= "This text should start at left bottom of the page and continue upwards through its left margin"
	) 
print(banner)

a,x,y=270,208,5 # right margin banner using .set_xy(x,y) and .cell() 
pdf.set_auto_page_break(False)
pdf.set_xy(x,y)
pdf.set_line_width(0.25)
with pdf.rotation(angle=a, x=x, y=y):
pdf.cell(
	text= "This text should start at right top of the page and continue downwards through its right margin",
	border=1, 
	h=10, # box height 10 mm
	# w=0 # I guess in rotation context this does not work as expected:
	# how can I extend text box downwards, far beyond text end, near to page bottom?
	) 

pdf.output("fpdf-lateral-banner-text.pdf")
```

Two more doubts:
- Does `w` parameter become useless when `.cell()` is used in a `rotation()` context? (see comment before `pdf.output`)
- Should I `set_auto_page_break(True)` again before continuing adding other stuff to my pdf? (not quite understood how that setting affects in this example)

Thanks a lot again!!

**Reply by gmischler** [reply]
The rotation context doesn't change anything about the functionality of its contents. It simply transforms the results to a different place and orientation. So whatever works elsewhere will work within a rotation context exactly the same. 

Your example only needs `.set_auto_page_break(False)` because your geometric arrangement makes it necessary to write text below the bottom page margin (you could also set the bottom margin to 0 to avoid the problem). If you later want to add normal horizontal running text that is supposed to cause page breaks, then yes, you'll have to enable those again.

If your box always has the same size (possibly with different text on different pages), then a `FlexTemplate()` might really be the simpler solution, even though you'll have to explicitly add the borders to it.

**Reply by abubelinha** [reply]
Thanks for your explanations again.

Regarding `w=0` option in `.cell()`[docstring](https://py-pdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.cell), I understand it extends the cell (and its border?) up to the right margin.
As my cell is rotated 270 degrees, I would expect `w=0` to extend cell's border up to the bottom margin.

But in my above code, it doesn't work like that: uncommenting `w=0` draws the cell border upwards (between text and page top) instead going downwards (around text and beyond up to bottom).
At least in my machine (fpdf2 2.7.8, Python 3.8, Windows 7)

I'd just prefer to have equal sized boxes in all pages because -as you point out- text length may differ.
Not a big deal, though.  That's a good motivation to learn how `FlexTemplate()` works.

**Reply by gmischler** [reply]
As explained just above, all that `rotation()` does is to wrap its contents in a transformation matrix. The `cell()` code is not aware in any way that its results are going to be moved around. So yes, `w=0` is of limited usefulness in this situation. You'll have to figure out a suitable value for that yourself.

Or you could really just use `FlexTemplate()`. It is designed specifically for this type of use case. Think of it as a rubber stamp with configurable text.
